src/main/java/com/redhat/devtools/lsp4ij/features/documentSymbol/LSPDocumentSymbolStructureViewModel.java

```diff
@@ -0,0 +1,227 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at https://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * CppCXY
+ ******************************************************************************/
+package com.redhat.devtools.lsp4ij.features.documentSymbol;
+
+import com.intellij.ide.structureView.StructureViewModel;
+import com.intellij.ide.structureView.StructureViewModelBase;
+import com.intellij.ide.structureView.StructureViewTreeElement;
+import com.intellij.ide.util.treeView.smartTree.TreeElement;
+import com.intellij.navigation.ItemPresentation;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.fileEditor.OpenFileDescriptor;
+import com.intellij.openapi.progress.ProcessCanceledException;
+import com.intellij.openapi.util.NlsSafe;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import com.intellij.psi.PsiTreeChangeAdapter;
+import com.intellij.psi.PsiTreeChangeEvent;
+import com.intellij.util.ui.update.MergingUpdateQueue;
+import com.intellij.util.ui.update.Update;
+import com.redhat.devtools.lsp4ij.LSPFileSupport;
+import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import org.eclipse.lsp4j.DocumentSymbolParams;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+
+import static com.redhat.devtools.lsp4ij.internal.CompletableFutures.isDoneNormally;
+import static com.redhat.devtools.lsp4ij.internal.CompletableFutures.waitUntilDone;
+
+/**
+ * LSP document symbol structure view model.
+ */
+public class LSPDocumentSymbolStructureViewModel extends StructureViewModelBase implements StructureViewModel.ElementInfoProvider {
+    private final PsiFile psiFile;
+    // Update queue to avoid multiple updates
+    private final MergingUpdateQueue updateQueue;
+
+    public LSPDocumentSymbolStructureViewModel(@NotNull PsiFile psiFile, @Nullable Editor editor) {
+        super(psiFile, editor, new LSPFileStructureViewElement(psiFile));
+        this.psiFile = psiFile;
+        this.updateQueue = new MergingUpdateQueue("lsp.documentSymbol", 300, true, null);
+        // Register psi tree change listener
+        registerPsiTreeChangeListener();
```

Comment:
You add a listener, but you remove it never. I think you should override the dispose method of the LSPDocumentSymbolStructureViewModel and remove the listener here.

Reply:
> You add a listener, but you remove it never. I think you should override the dispose method of the LSPDocumentSymbolStructureViewModel and remove the listener here.

the signature like this:
![image](https://github.com/user-attachments/assets/9ed04f13-7782-434e-9a79-9adc354489dd)
I've already passed in the parent, so I think explicit dispose shouldn't be necessary.

Reply:
+1, I wil check debug if it is working. I think you need to override dipose to:

* cancel the LSPDocumentSymbolSupport.cancel() 
* cancel updateQueue like (not sure ?)

```java
updateQueue.cancelAllUpdates()
updateQueue.deactivate()
```